pkg/ifaces/watcher.go

```diff
@@ -16,59 +21,75 @@ type Watcher struct {
+	linkSubscriberAt func(ns netns.NsHandle, ch chan<- netlink.LinkUpdate, done <-chan struct{}) error
+	mutex            *sync.Mutex
 }
 
 func NewWatcher(bufLen int) *Watcher {
 	return &Watcher{
-		bufLen:         bufLen,
-		current:        map[Interface]struct{}{},
-		interfaces:     netInterfaces,
-		linkSubscriber: netlink.LinkSubscribe,
+		bufLen:           bufLen,
+		current:          map[Interface]struct{}{},
+		interfaces:       netInterfaces,
+		linkSubscriberAt: netlink.LinkSubscribeAt,
+		mutex:            &sync.Mutex{},
 	}
 }
 
 func (w *Watcher) Subscribe(ctx context.Context) (<-chan Event, error) {
 	out := make(chan Event, w.bufLen)
 
-	go w.sendUpdates(ctx, out)
+	nsHandles, err := getNetNSHandles(ctx)
+	if err != nil {
+		go w.sendUpdates(ctx, netns.None(), out)
+	} else {
+		for _, nsh := range nsHandles {
+			nsHandle := nsh
+			go w.sendUpdates(ctx, nsHandle, out)
+		}
+	}
 
 	return out, nil
 }
 
-func (w *Watcher) sendUpdates(ctx context.Context, out chan Event) {
+func (w *Watcher) sendUpdates(ctx context.Context, netnsHandle netns.NsHandle, out chan Event) {
 	log := logrus.WithField("component", "ifaces.Watcher")
-
 	// subscribe for interface events
 	links := make(chan netlink.LinkUpdate)
-	if err := w.linkSubscriber(links, ctx.Done()); err != nil {
-		log.WithError(err).Error("can't subscribe to links")
+	log.WithField("netns", netnsHandle.String()).Debug("linkSubscribe to receive links update")
+	if err := w.linkSubscriberAt(netnsHandle, links, ctx.Done()); err != nil {
+		log.WithError(err).Errorf("can't subscribe to links netns %s", netnsHandle.String())
 		return
 	}
 
 	// before sending netlink updates, send all the existing interfaces at the moment of starting
 	// the Watcher
-	if names, err := w.interfaces(); err != nil {
-		log.WithError(err).Error("can't fetch network interfaces. You might be missing flows")
-	} else {
-		for _, name := range names {
-			w.current[name] = struct{}{}
-			out <- Event{Type: EventAdded, Interface: name}
+	if netnsHandle.Equal(netns.None()) {
+		if names, err := w.interfaces(); err != nil {
+			log.WithError(err).Error("can't fetch network interfaces. You might be missing flows")
+		} else {
+			for _, name := range names {
+				iface := Interface{Name: name.Name, Index: name.Index, NetNS: netnsHandle}
+				w.mutex.Lock()
+				w.current[iface] = struct{}{}
+				w.mutex.Unlock()
+				out <- Event{Type: EventAdded, Interface: iface}
+			}
 		}
 	}
-
 	for link := range links {
 		attrs := link.Attrs()
 		if attrs == nil {
 			log.WithField("link", link).Debug("received link update without attributes. Ignoring")
 			continue
 		}
-		iface := Interface{Name: attrs.Name, Index: attrs.Index}
-		if link.Flags&(syscall.IFF_UP|syscall.IFF_RUNNING) != 0 {
+		iface := Interface{Name: attrs.Name, Index: attrs.Index, NetNS: netnsHandle}
+		w.mutex.Lock()
+		if link.Flags&(syscall.IFF_UP|syscall.IFF_RUNNING) != 0 && attrs.OperState == netlink.OperUp {
```

Comment:
For the record `attrs.OperState == netlink.OperUp` is needed to avoid calling `EventAdded` too soon since it will come up with `OperState` set to `Down` first.